### PR TITLE
[RISCV] Place mergeable small read only data into srodata section

### PR DESCRIPTION
**Pull request by kito-cheng**
Small mergeable read only data was place on the sdata before, but it also means it lose the mergeable property, which means lose some code size optimization opportunity during link time.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp [hunk 1]
```diff
@@ -126,8 +136,20 @@ bool RISCVELFTargetObjectFile::isConstantInSmallSection(
 MCSection *RISCVELFTargetObjectFile::getSectionForConstant(
     const DataLayout &DL, SectionKind Kind, const Constant *C,
     Align &Alignment) const {
-  if (isConstantInSmallSection(DL, C))
-    return SmallDataSection;
+  if (isConstantInSmallSection(DL, C)) {
+    if (Kind.isMergeableConst4())
+      return SmallROData4Section;
+    else if (Kind.isMergeableConst8())
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetObjectFile.cpp:142]: No else after return

**Comment by topperc** [comment]
Does gcc do something similar?

**Comment by kito-cheng** [comment]
> Does gcc do something similar?

Yeah, here is the reference: https://github.com/gcc-mirror/gcc/blob/master/gcc/config/riscv/riscv.cc#L6177-L6189

**Comment by MaskRay** [comment]
I am still interested in changing the default limit from 8 to 0. Is there interest for GCC to do so?

**Comment by kito-cheng** [comment]
> I am still interested in changing the default limit from 8 to 0. Is there interest for GCC to do so?

From the benchmarking aspect, it's hard to say I am interesting on change the default value honestly, since I didn't got the point about the disadvantage when the GP relaxation is turn off, but it did help when GP relaxation turn on, however I may missed something, so I am happy to discuss this more, and change my mind IF we gather enough arguments :)

|                             | w/ GP relax                        | w/o GP relax |
| ---------------------- | ----------------------------------- | ------------------- |
| Default limit 0   | Less chance on GP relax `*1` | No difference?`*3` |
| Default limit 8   | More chance on GP relax `*2` | No difference |


`*1`: because the range is limited, so same range cover more symbol have a higher winning rate.
`*2`: Compare to `Default limit 0`
`*3`: I guess I may missed something here, so need your input here.

**Comment by stephan-gh** [comment]
I believe this PR might fix the problem I reported on the LLVM security bug tracker back in January, since I was unsure about the security implications: https://bugs.chromium.org/p/llvm/issues/detail?id=61 It was decided that this is better handled publicly.

Aside from not being able to merge those constants the current behavior of placing constants in .sdata by default is also non-ideal because it makes it impossible to differentiate between small **read-only** constants and actual small **mutable** global variables. This is necessary if you want to favor security hardening (by mapping the constants read-only at runtime) over the slight performance/size benefit of linker relaxation. For example, [in the Linux kernel a custom linker script is used to place .srodata next to .rodata](https://github.com/torvalds/linux/blob/v6.7/arch/riscv/kernel/vmlinux.lds.S#L120-L125) rather than merging it into the mutable .data. This works only with GCC at the moment because LLVM does not generate .srodata.

In my report linked above, I suggested setting `-msmall-data-limit` to 0 by default as a workaround but these changes look like they would fix the actual problem. :)

**Review by topperc — APPROVED** [review]
This LGTM. @MaskRay does it look good to you? I think independent of the default small data limit we still need to fix this when there is a small data limit.

**Review by ilovepi — APPROVED** [review]
LGTM from my perspective, and I agree this seems likely to address the issue raised in https://bugs.chromium.org/p/llvm/issues/detail?id=61 . In Fuchsia, we set `-msmall-data-limit=0` everywhere, in part due to that, but that isn't the only concern. We've set that for a long time, in part due to our use of the Shadow Call Stack, and some interactions w/ small data limit were not desirable. I'm not completely sure if those still exists, though. So I agree w/ @MaskRay 's suggestion about choosing a different default, though that is an orthogonal question.

**Review by petrhosek — APPROVED** [review]
LGTM this is beneficial for our embedded/baremetal use cases.

**Comment by kito-cheng** [comment]
Gonna merge, I believe it's an improvement, and default value of that is separated issue :)

**Pull request merged**